Hide inaccessible navbar breadcrumbs in enhanced nav

miguelaenlle opened the pull request:
Closes #11553. 

Currently, the "Institutions", "Institution Name" (e.g. "UIUC"), and "Courses" are visible to users when they have `institution_administrator` privileges. However, the pages these links redirect to, 

- `/pl/administrator/institutions`
- `/pl/administrator/institution/${resLocals.institution.id}`
- `/pl/administrator/institution/${resLocals.institution.id}/courses`

Require global administrator access:

https://github.com/PrairieLearn/PrairieLearn/blob/590b47abdb520f1a1eb1027f03d35bcd01ae43e8/apps/prairielearn/src/server.ts#L1873

https://github.com/PrairieLearn/PrairieLearn/blob/590b47abdb520f1a1eb1027f03d35bcd01ae43e8/apps/prairielearn/src/middlewares/authzIsAdministrator.ts#L1-L10

Therefore, some users with `institution_administrator` privileges won't have access to these breadcrumb links.

This PR adjusts the enhanced navigation navbar to only display those tabs when the user has global administrator access.

Global administrator view:

![Screenshot 2025-03-16 at 8 18 57 PM](https://github.com/user-attachments/assets/cce3c849-4c5e-4b8f-86f2-82dc5d557aa0)

Non-global administrator view:

![image](https://github.com/user-attachments/assets/4c112f21-df10-4607-8115-dd8fd93b9c9a)

Questions:
- Home > Course Name is the most content the navbar will display for users who are not global administrators. Should the navbar also display the current course instance, and possibly the current subpage (e.g. Questions, Assessments)? This would populate the navbar, but be redundant information already displayed on the sidebar.

nwalters512 commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -671,29 +671,28 @@ function NavbarButtons({
     allNavbarButtons.push({ text: 'Global Admin', href: '/pl/administrator/admins' });
-  }
-
-  if (resLocals.is_institution_administrator && resLocals.institution) {
```
We do actually have two different types of administrators: global admins (who can see everything) and institution admins (admins within a given institution). I do think it'd be good for institution admins to retain the ability to see the institution in the nav bar and have that as a link that jumps to the list of courses for the institution (e.g. `/pl/institution/1/admin/courses`).

That is:

- If someone is a global admin, they'll see the full `Home -> Institutions -> INSTITUTION_NAME -> Courses -> COURSENAME`; "institutions", "INSTITUTION_NAME", and "Courses" are links to the global admin pages for those things.
- Otherwise, if someone is an institution admin, they'll see `Home -> INSTITUTION_NAME -> Courses -> COURSE_NAME`; "INSTITUTION_NAME" and "Courses" are links to the _institution admin_ pages for those things.

How does all this sound?

miguelaenlle replied:
That sounds good, @nwalters512 ! Thank you for the info. I implemented this functionality. 

Global admin view:
![image](https://github.com/user-attachments/assets/d98be3af-9718-45ab-9829-8fda40b3f1a4)

Institution admin view:
![image](https://github.com/user-attachments/assets/6570cf6c-d432-44a4-ae17-21d4726b3c03)

- "INSTITUTION_NAME" breadcrumb redirects to `/pl/institution/1/admin/admins`
- "Courses" breadcrumb redirects to `/pl/institution/1/admin/courses`

Regular user view:
![image](https://github.com/user-attachments/assets/3acda4f7-aeec-4a4f-98cb-2ad4c6df4c9d)

nwalters512 replied:
This logic looks better now! I'll close this conversation, we'll continue the threads elsewhere about the actual authorization logic.

nwalters512 reviewed (changes requested):
Testing of this PR revealed two bugs in the implementation of `res.locals.is_institution_administrator`:

- The `LEFT JOIN` here should be a `JOIN`; otherwise it will always return true: https://github.com/PrairieLearn/PrairieLearn/blob/6353b381c99fd1fe0181d47ff4e63545ef6f3939/apps/prairielearn/src/lib/authn.sql#L21-L34 In fact, this can be simplified such that it doesn't even join on `institutions` and instead just does something like this (untested):
  ```sql
  SELECT
    EXISTS (
      SELECT
        1
      FROM
        institution_administrators
      WHERE
        user_id = $user_id
        AND institution_id = $institution_id
    );
  ```
- This check should use `res.locals.is_administrator` so that it properly reflects the "Turn on/off administrator access" setting in the user menu in the upper right: https://github.com/PrairieLearn/PrairieLearn/blob/6353b381c99fd1fe0181d47ff4e63545ef6f3939/apps/prairielearn/src/lib/authn.ts#L122

miguelaenlle commented:
Thank you for the feedback, @nwalters512 . I adjusted the SQL query and the `is_institution_administrator` logic. 

https://github.com/PrairieLearn/PrairieLearn/blob/7909215897066d65341499b0ffe6cf68ede0e1ea/apps/prairielearn/src/lib/authn.ts#L122-L135

I added the additional `res.locals.authn_is_administrator && !res.locals.access_as_administrator` check, since institution admin pages aren't accessible if an administrator doesn't access as one: 

https://github.com/PrairieLearn/PrairieLearn/blob/a55dae87b4f786ea80a76f45676dc515e82bdd1e/apps/prairielearn/src/ee/lib/selectAndAuthz.ts#L29-L35

Here were the results from testing the enhanced nav breadcrumbs:

Global administrator:

![image](https://github.com/user-attachments/assets/9c26f1a2-fca5-4aec-ac9c-6226cd1035fa)

Global administrator not accessing as administrator:

![image](https://github.com/user-attachments/assets/75333df8-c47e-45b6-85be-c348f49ee2c9)

Institution administrator:

![image](https://github.com/user-attachments/assets/0b40903f-6ba9-48ee-ad67-c79ef1a82b68)

Standard user:

![image](https://github.com/user-attachments/assets/0a313551-176d-4016-a025-833ea9c79de2)

miguelaenlle commented on apps/prairielearn/src/lib/authn.ts:
```diff
@@ -118,16 +119,23 @@ export async function loadUser(
+    // Institution admin pages are inaccessible if the user is an administrator not accessing as one.
+    if (res.locals.authn_is_administrator && !res.locals.access_as_administrator) {
+      return false;
+    }
```
Sounds good; I replaced this check.

As per our discussion on Slack, I'll open a separate PR to adjust `selectAndAuthzInstitutionAsAdmin` to provide institution admin page access to global admins with admin access off who are explicitly added as institution admins.

miguelaenlle replied:
This PR is now open at https://github.com/PrairieLearn/PrairieLearn/pull/11593.

nwalters512 reviewed (approved):
I tested this locally, it's looking good!
